### PR TITLE
Adds compat checker for versioning and compatibility

### LINKED ISSUE #2639: Add WooCommerce compatibility and versioning
## User story

As a user, 

* I'd like to be notified when I am using a WooCommerce version that will no longer be supported soon.
* I'd like my plugin to not brick my site if I am using a WooCommerce version that is not supported.

### PR DESCRIPTION
**Pull request by ibndawood**
### Changes proposed in this Pull Request:

Closes #2639.

This PR uses the `compat-checker` library from the [WooCommerce Grow repo](https://github.com/woocommerce/grow) to run compatibility and versioning checks during the plugin initialisation.

### Checks:

* [x] Does your code follow the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Screenshots:

Screenshots are for reference. **WooCommerce Brands** will be replaced by **Facebook for WooCommerce**.

#### WooCommerce not installed
image

#### WooCommerce not activated

image

#### WooCommerce not compatible

image

#### WooCommerce not tested

image

#### Pre-requisites

The PR requires deleting transients. A plugin like [Transients Manager](https://wordpress.org/plugins/transients-manager/) is recommended to delete the transients between tests.

#### WooCommerce activation tests

With the above code snippets in place, run the following tests. The instructions are for Facebook for WooCommerce but can be used for other WooGrow Extensions.

1. Install and activate Facebook for WooCommerce without installing WooCommerce.
2. Deactivate WooCommerce.
3. Verify the notice `Facebook for WooCommerce requires WooCommerce to be installed and activated. Please install WooCommerce.` is displayed.
4. Install WooCommerce without activating.
5. Verify the notice `Facebook for WooCommerce requires WooCommerce to be activated. Please activate WooCommerce.` is displayed.

#### WooCommerce versioning tests
⚠️  **Important note**: You will need to erase the transient `_transient_wc_grow_compat_checker_[EXTENSION].php[VERSION` after each change!!

1. Install and activate WooCommerce.
2. Install and activate Facebook for WooCommerce.
3. Modify the `WC requires at least` header field and set it to a version higher than the active WooCommerce version. Let's say the `WC requires at least` is set to `8.2` and the active WooCommerce version is `8.0.2`.
4. Verify the notice `Facebook for WooCommerce requires WooCommerce version 8.2 or higher. Please update WooCommerce to the latest version, or download the minimum required version` is displayed.
5. Modify the `WC tested up to` and set it to a lower version than the active WooCommerce version. Let's say the `WC tested up to` is set to `7.8`, the active WooCommerce version is `8.0.2` and Facebook for WooCommerce version is `3.0.34`.
6. Verify the warning notice `Facebook for WooCommerce - 3.0.34 is untested with WooCommerce 8.0.2` is displayed.

#### L2 support recommendation notice

1. Download the latest minus three WooCommerce version and activate it. Let's say `7.6`.
2. Verify the warning notice `Heads up! Facebook for WooCommerce will soon discontinue support for WooCommerce 7.6. Please update WooCommerce to take advantage of the latest updates and features.`

#### WordPress versioning tests
⚠️  **Important note**: You will need to erase the transient `_transient_wc_grow_compat_checker_[EXTENSION].php[VERSION` after each change!!

1. Modify the `Tested up to` and set it to a lower version than the current WordPress version. Let's say the `Tested up to` is set to `6.2`, the active WordPress version is `6.3`, and the Facebook for WooCommerce version is `3.0.34`.
2. Verify the warning notice `Facebook for WooCommerce - 3.0.34 is untested with WordPress 6.3` is displayed.

### Additional Details

To delete a transient using Transients Manager, navigate to Tools > Transients, search for `wc_grow` and delete the transient.

### Changelog entry

> Add - Versioning and compatibility checks to implement support policy.

### CODE REVIEW

File: class-wc-facebookcommerce.php [hunk 1]
```diff
@@ -9,8 +9,7 @@
  * @package FacebookCommerce
  */
 
-require_once __DIR__ . '/vendor/autoload.php';
-require_once __DIR__ . '/includes/fbutils.php';
+ require_once __DIR__ . '/includes/fbutils.php';
```
Comment (rawdreeg) [inline class-wc-facebookcommerce.php:12]:
```suggestion
require_once __DIR__ . '/includes/fbutils.php';
```

File: facebook-for-woocommerce.php [hunk 2]
```diff
@@ -127,7 +129,7 @@ public function __wakeup() {
 	 */
 	public function init_plugin() {
 
-		if ( ! $this->plugins_compatible() ) {
+		if ( ! Checker::instance()->is_compatible( __FILE__, self::PLUGIN_VERSION ) || ! $this->plugins_compatible() ) {
```
Comment (rawdreeg) [inline facebook-for-woocommerce.php:132]: Shouldn't Checker::instance()->is_compatible( __FILE__, self::PLUGIN_VERSION ) move inside plugins_compatible ?

Comment (ibndawood) [reply facebook-for-woocommerce.php:132]: This is a good catch, @rawdreeg. My approach was not to disturb existing checks. However, in this case, we can remove the `plugins_compatible` method and replace it with `Checker::instance()->is_compatible( FILE, self::PLUGIN_VERSION )` alone.

Comment (ibndawood) [reply facebook-for-woocommerce.php]: I've implemented the above suggestion in this commit: https://github.com/woocommerce/facebook-for-woocommerce/pull/2640/commits/8eefaf340d5c5f310805b7b0dc07f9c40e3e46fc

**Comment by ibndawood** [comment]
Thank you, @rawdreeg, for taking the time to review the PR. I've pushed a couple of commits addressing your comments. Please let me know your thoughts.

File: facebook-for-woocommerce.php [hunk 3]
```diff
@@ -127,7 +129,7 @@ public function __wakeup() {
 	 */
 	public function init_plugin() {
 
-		if ( ! $this->plugins_compatible() ) {
+		if ( ! Checker::instance()->is_compatible( __FILE__, self::PLUGIN_VERSION ) ) {
```
Comment (rawdreeg) [inline facebook-for-woocommerce.php:132]: This probably means we can remove plugins_compatible as this is a private method and it's no longer used :-)

**Review by rawdreeg — APPROVED** [review]
Looks good, Thanks @ibndawood . I added a suggestion to remove the plugins_compatible method. but that's left up to you :-)

**Pull request merged**